### PR TITLE
Fixed with Enable Lint:DrawAllocation #10488

### LINKED ISSUE #10488: Enable Lint: `DrawAllocation`
We perform lint checking to ensure code correctness and improve performance, but many of the system lint checks have not been enabled.

The goal of this issue is to enable the lint check: `DrawAllocation`

Line: https://github.com/ankidroid/Anki-Android/blob/6260cd6e25f67f331833bbd0981f2cda5a426dde/lint-release.xml#L208

To do this: 

* Change `ignore` to `fatal`
* Move the line in the XML to be somewhere near the top of the file (before the large block of `ignore` elements. 
  * The exact position is up to you, but if there's no comment associated with the check, a large nearly alphabetical block of `fatal` warnings seems sensible
* Run `gradlew lint` in the Android Studio terminal and fix the errors which occur.
  * `@Suppress(str)` (Kotlin) or `@SuppressWarnings(str)`  (Java) the warnings if they are valid. Android Studio's quickfixes menu will offer options to suppress
    * Use your judgment, but aim to suppress as close to the error as possible (line/method) unless the class specifically has a reason why it will violate a lint check many times
  * If it was an error that we missed, fix it.
* Submit a Pull Request

### PR DESCRIPTION
**Pull request by Bhawna1203**
## Purpose / Description
Enabled lint check for "DrawAllocation", changed "ignore" to "fatal" and took the line above.

## Fixes
Fixes #10488 

## Approach
I just go into the Project file section and go into the file  lint-release.xml and in line 208 i change the issue tag severity into fatal .

## How Has This Been Tested?
Ran ./gradlew lint 
Everything runs fine so i create the pull request

## Learning (optional, can help others)
I read the codebase of this project and learn a lot .

### CODE REVIEW

**Comment by Bhawna1203** [comment]
Please check @david-allison .
Thanks

**Comment by viciousAegis** [comment]
Hey! It would be great if you could fill up the Pull Request template, and especially add a link to the issue that this is fixing under the `Fixes` heading.

Apart from that, I see that the commit changes the `README` file? 
Please correct me if I'm wrong, but from my understanding of the issue I don't think changes should me made to this file.

File: lint-release.xml [hunk 1]
```diff
@@ -205,7 +205,7 @@
     <issue id="TrustAllX509TrustManager" severity="ignore" />
     <issue id="InvalidImeActionId" severity="ignore" />
     <issue id="InvalidPackage" severity="ignore" />
-    <issue id="DrawAllocation" severity="ignore" />
+    <issue id="DrawAllocation" severity="fatal" />
```
Comment (viciousAegis) [inline lint-release.xml:208]: Since you've changed the severity to `fatal`, it would be great if you moved the entire statement a few lines above this big block of `ignore` issues along with the other `fatal` issues. You can place it along with those issues as you see fit.

Comment (viciousAegis) [inline lint-release.xml:208]: Since you've changed the severity to `fatal`, it would be great if you moved the entire statement a few lines above this big block of `ignore` issues along with the other `fatal` issues. You can place it along with those issues as you see fit.  see https://github.com/ankidroid/Anki-Android/issues/10488#issue-1166759351:  

> Move the line in the XML to be somewhere near the top of the file (before the large block of ignore elements.
> The exact position is up to you, but if there's no comment associated with the check, a large nearly alphabetical block of fatal warnings seems sensible

File: README.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <p align="center">
 <img src="docs/graphics/logos/banner_readme.png"/>
 </p>
-
+<p>
```
Comment (david-allison) [inline README.md:4]: this file shouldn't have been changed

**Review by david-allison — CHANGES_REQUESTED** [review]
Something is wrong with `gradlew lint`. Try it again. It should have been fatal

```
> Task :AnkiDroid:lintAmazonDebug FAILED
C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\anki\stats\ChartView.kt:52: Error: Avoid object allocations during draw/layout operations (preallocate and reuse instead) [DrawAllocation]
            val paint = Paint(Paint.LINEAR_TEXT_FLAG)
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\anki\stats\ChartView.kt:55: Error: Avoid object allocations during draw/layout operations (preallocate and reuse instead) [DrawAllocation]
            val g = GraphicsWrap(canvas, paint)
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\anki\stats\ChartView.kt:56: Error: Avoid object allocations during draw/layout operations (preallocate and reuse instead) [DrawAllocation]
            val field = Rect()
                        ~~~~~~
3 errors, 0 warnings


```

File: lint-release.xml (hunk 1 above)
Comment (Bhawna1203) [reply lint-release.xml:208]: Thanks for helping me i will again create pull request tomorrow with fixing this issue.

File: README.md (hunk 2 above)
Comment (Bhawna1203) [reply README.md:4]: I did this by mistake .I will change this tomorrow sir.Thanks for helping me sir.

File: lint-release.xml (hunk 1 above)
Comment (viciousAegis) [reply lint-release.xml:208]: No, you don't have to make another pull request, and don't close this one. You can just add these in another commit, and push (possibly force-push) them to the branch you have created. If you don't have much knowledge about these git commands, don't worry, I was there not even a month ago, but some googling can help you here. 

TLDR: Don't make another PR, add commits in this one only.

Comment (Bhawna1203) [reply lint-release.xml:208]: Okk sir .Thanks for helping .Actually this one is my first big codebase code so i am facing issues.

File: README.md (hunk 2 above)
Comment (Bhawna1203) [reply README.md:4]: Actually here the paragraph tag is not opened before closing so it gives error so i think to fix it. 
Sorry for that sir

File: lint-release.xml (hunk 1 above)
Comment (Bhawna1203) [reply lint-release.xml:208]: Now please check i did some changes .If issue again persist then let me know

Comment (viciousAegis) [reply lint-release.xml:208]: @Bhawna1203 I can only see one commit. Did you push the commits to your remote branch?

Comment (Bhawna1203) [reply lint-release.xml:208]: Yess

**Comment by david-allison** [comment]
Still waiting for feedback to be addressed

**Comment by Bhawna1203** [comment]
@david-allison sir, Please check i do some commit .
 If it works then let me know Or need to change something then again please let me know

File: README.md (hunk 2 above)
Comment (mikehardy) [reply README.md:4]: I am looking at all commits in this file. I still see this change to README in here. README may indeed have problems but they are unrelated to enabling a lint check, thus no changes to README should be mixed in to this PR.
This change must be removed from the PR

File: lint-release.xml (hunk 1 above)
Comment (mikehardy) [reply lint-release.xml:208]: I am not sure what changes exist on your local system, but they have not been pushed to github.
This change from ignore to fatal is still buried in the ignore block of checks and should be moved up per previous comments

**Review by mikehardy — CHANGES_REQUESTED** [review]
still needs the changes requested to be visible on github.com so we can see them

**Comment by Bhawna1203** [comment]
Okk @mikehardy sir i will do the changes and let you know when i completed the same.Thanks for helping and continuous support.

**Comment by Bhawna1203** [comment]
@david-allison sir after completing all the changes as instructed in the issue page .I run gradlew-lint and getting some warning .Could i commit these changes or need to change something?
Screenshot from 2022-03-17 18-29-36

**Comment by david-allison** [comment]
Those are fine

**Comment by Bhawna1203** [comment]
So @david-allison sir ,Can i now commits the changes for review

**Comment by david-allison** [comment]
No need to ask! I can't review what I can't see

**Comment by Bhawna1203** [comment]
Sir, After solving  gradlew-lint issue . Sir i am getting this issue. Do i need to solve these issues to . 
Could you please help there as a teacher or mentor.


Screenshot from 2022-03-19 15-17-08

**Comment by david-allison** [comment]
That doesn't look related. 

If you were able to build AnkiDroid before, it's probably a bug that you introduced, and need to fix

If you weren't, you should focus on getting AnkiDroid to build before taking on issues.

**Comment by Bhawna1203** [comment]
@david-allison sir Sorry for inconvenience. I solved this issue earlier. But due to updation in Ubantu today when i open this project again in my Android Studio. It gives me issue that i shared above. I tried to solve issue but i am unable to solve about 6 hours of constant efforts.
Then again i cloned the project and solved the issue  https://github.com/ankidroid/Anki-Android/issues/10488 again  and its solved. so i need to put new PR, i tested with gradlew lint and also it succeed .I added the screenshot here too.I am embarrassed for this convenience Hope you understand.  
So can I push that new cloned version ? and add new PR?

Screenshot from 2022-03-19 17-40-24
Screenshot from 2022-03-19 17-26-21

**Comment by david-allison** [comment]
Use this branch: `[Bhawna1203:bhawna-changes](https://github.com/Bhawna1203/Anki-Android/tree/bhawna-changes)`. Don't make a new PR.

There's no time pressure. First contributions typically take a lot longer than expected, but momentum builds up. There's no need to feel embarrassed, and no need to say how long something took, unless you want to share it. 

I've definitely had pull requests that are only a few lines that took hours or days to write. It's fine.

**Comment by Bhawna1203** [comment]
@david-allison sir check i am done with the changes. Also i am not getting any issue on gradlew lint 
Screenshot from 2022-03-20 15-54-22

File: README.md [hunk 3]
```diff
@@ -14,6 +14,7 @@
 <a href="https://github.com/ankidroid/Anki-Android/blob/main/COPYING"><img src="https://img.shields.io/github/license/ankidroid/Anki-Android" alt="license"/></a>
 </p>
 
+
```
Comment (david-allison) [inline README.md:17]: Please revert all changes to this file

File: AnkiDroid/src/main/java/com/ichi2/anki/stats/ChartView.kt [hunk 4]
```diff
@@ -45,6 +46,7 @@ class ChartView : View {
         setWillNotDraw(false)
     }
 
+    @SuppressLint("DrawAllocation")
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/stats/ChartView.kt:49]: We should be fixing the errors, not ignoring them. They're actual problems

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/stats/ChartView.kt (hunk 4 above)
Comment (Bhawna1203) [reply AnkiDroid/src/main/java/com/ichi2/anki/stats/ChartView.kt:49]: Okk sir

File: README.md (hunk 3 above)
Comment (Bhawna1203) [reply README.md:17]: okkk sir

**Comment by mikehardy** [comment]
I'm going to change this to draft so it doesn't show up in the scans I do for actionable PRs yet - it's easy to toggle it back and forth though - just a button on Github's Web UI to set it back to ready when it's ready

File: AnkiDroid/src/main/java/com/ichi2/anki/stats/ChartView.kt (hunk 4 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/stats/ChartView.kt:49]: > DrawAllocation:
> 
> Memory allocations within drawing code
> 
> Explanation:
> 
> Looks for memory allocations within drawing code. You should avoid allocating objects during a drawing or layout operation. These are called frequently, so a smooth UI can be interrupted by garbage collection pauses caused by the object allocations.
> 
> The way this is generally handled is to allocate the needed objects up front and to reuse them for each drawing operation.
> 
> Some methods allocate memory on your behalf (such as Bitmap.create), and these should be handled in the same way.

Additional information: https://stackoverflow.com/a/27719578/
